File: editor-ace/src/main/java/com/raredev/vcspace/editor/AceCodeEditor.java
package com.raredev.vcspace.editor;

import android.content.Context;
import android.util.AttributeSet;
import android.view.KeyEvent;
import android.view.inputmethod.BaseInputConnection;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputConnection;
import android.webkit.WebView;
import androidx.javascriptengine.JavaScriptSandbox;
import com.google.common.util.concurrent.ListenableFuture;
import com.raredev.vcspace.editor.ace.OnEditorLoadedListener;
import com.raredev.vcspace.editor.ace.OnJsReturnValueCallback;
import com.raredev.vcspace.editor.ace.VCSpaceEditorLanguage;
import com.raredev.vcspace.editor.ace.VCSpaceEditorTheme;
import com.raredev.vcspace.editor.ace.VCSpaceWebChromeClient;
import com.raredev.vcspace.editor.ace.VCSpaceWebInterface;
import com.raredev.vcspace.editor.ace.VCSpaceWebViewClient;
import com.raredev.vcspace.editor.ace.options.EditorOptions;
import java.io.File;

public class AceCodeEditor extends WebView {

  private float x, y;
  private ListenableFuture<JavaScriptSandbox> jsSandbox;

  public OnEditorLoadedListener onEditorLoadedListener;
  private VCSpaceWebInterface webInterface;
  private EditorOptions options;

  private boolean modified;
  private File file;

  public AceCodeEditor(Context context) {
    this(context, null);
  }

  public AceCodeEditor(Context context, AttributeSet attrs) {
    this(context, attrs, 0);
  }

  public AceCodeEditor(Context context, AttributeSet attrs, int defStyleAttr) {
    this(context, attrs, defStyleAttr, 0);
  }

  public AceCodeEditor(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
    super(context, attrs, defStyleAttr, defStyleRes);
    jsSandbox = JavaScriptSandbox.createConnectedInstanceAsync(context.getApplicationContext());
    options = EditorOptions.getInstance();

    // setOnTouchListener(onTouchListener());
    addJavascriptInterface(webInterface = new VCSpaceWebInterface(this), "VCSpace");
    setWebChromeClient(new VCSpaceWebChromeClient());
    setWebViewClient(new VCSpaceWebViewClient(this));

    getSettings().setJavaScriptEnabled(true);
    loadUrl("file:///android_asset/ace-editor/editor.html");
  }

  public void release() {
    try {
      jsSandbox.get().unbindService();
      jsSandbox.get().close();
    } catch (Exception e) {
      e.printStackTrace();
    }
    webInterface = null;
    destroy();
  }

  public void loadOptions() {
    var sb = new StringBuilder();
    sb.append("editor.setOptions({")
        .append("highlightActiveLine: ")
        .append(options.highlightActiveLine)
        .append(",")
        .append("highlightSelectedWord: ")
        .append(options.highlightSelectedWord)
        .append(",")
        .append("readOnly: ")
        .append(options.readOnly)
        .append(",")
        .append("enableAutoIndent: ")
        .append(options.enableAutoIndent)
        .append(",")
        .append("enableBasicAutocompletion: ")
        .append(options.enableBasicAutocompletion)
        .append(",")
        .append("enableLiveAutocompletion: ")
        .append(options.enableLiveAutocompletion)
        .append(",")
        .append("enableSnippets: ")
        .append(options.enableSnippets)
        .append(",")
        .append("hScrollBarAlwaysVisible: ")
        .append(options.hScrollBarAlwaysVisible)
        .append(",")
        .append("vScrollBarAlwaysVisible: ")
        .append(options.vScrollBarAlwaysVisible)
        .append(",")
        .append("highlightGutterLine: ")
        .append(options.highlightGutterLine)
        .append(",")
        .append("animatedScroll: ")
        .append(options.animatedScroll)
        .append(",")
        .append("showInvisibles: ")
        .append(options.showInvisibles)
        .append(",")
        .append("showPrintMargin: ")
        .append(options.showPrintMargin)
        .append(",")
        .append("fadeFoldWidgets: ")
        .append(options.fadeFoldWidgets)
        .append(",")
        .append("showFoldWidgets: ")
        .append(options.showFoldWidgets)
        .append(",")
        .append("showLineNumbers: ")
        .append(options.showLineNumbers)
        .append(",")
        .append("showGutter: ")
        .append(options.showGutter)
        .append(",")
        .append("displayIndentGuides: ")
        .append(options.displayIndentGuides)
        .append(",")
        .append("fixedWidthGutter: ")
        .append(options.fixedWidthGutter)
        .append(",")
        .append("useSvgGutterIcons: ")
        .append(options.useSvgGutterIcons)
        .append(",")
        .append("useSoftTabs: ")
        .append(options.useSoftTabs)
        .append(",")
        .append("enableEmmet: ")
        .append(options.enableEmmet)
        .append(",")
        .append("mergeUndoDeltas: ")
        .append(
            options.mergeUndoDeltas.equals(EditorOptions.MergeUndoDeltas.NEVER)
                ? false
                : options.mergeUndoDeltas.equals(EditorOptions.MergeUndoDeltas.TIMED)
                    ? true
                    : "'always'")
        .append(",")
        .append("selectionStyle: ")
        .append("\"" + options.selectionStyle.name().toLowerCase() + "\"")
        .append(",")
        .append("cursorStyle: ")
        .append("\"" + options.cursorStyle.name().toLowerCase() + "\"")
        .append(",")
        .append("theme: ")
        .append("\"ace/theme/" + options.theme.name().toLowerCase() + "\"")
        .append(",")
        .append("mode: ")
        .append("\"ace/mode/" + options.language.name().toLowerCase() + "\"")
        .append(",")
        .append("newLineMode: ")
        .append("\"" + options.newLineMode.name().toLowerCase() + "\"")
        .append(",")
        .append("foldStyle: ")
        .append("\"" + options.foldStyle.name().toLowerCase() + "\"")
        .append(",")
        .append("firstLineNumber: ")
        .append(options.firstLineNumber)
        .append(",")
        .append("fontSize: ")
        .append(options.fontSize)
        .append(",")
        .append("printMarginColumn: ")
        .append(options.printMarginColumn)
        .append(",")
        .append("tabSize: ")
        .append(options.tabSize)
        .append(",")
        .append("maxLines: ")
        .append(options.maxLines == 0 ? "undefined" : options.maxLines)
        .append(",")
        .append("minLines: ")
        .append(options.minLines == 0 ? "undefined" : options.minLines)
        .append(",")
        .append("scrollPastEnd: ")
        .append(options.scrollPastEnd)
        .append("})");

    loadJs(sb.toString());
  }

  public void setTheme(VCSpaceEditorTheme theme) {
    var sb = new StringBuilder();
    sb.append("editor.setTheme('ace/theme/").append(theme.name().toLowerCase()).append("')");
    loadJs(sb.toString());
  }

  public void setLanguage(VCSpaceEditorLanguage language) {
    var sb = new StringBuilder();
    sb.append("editor.session.setMode('ace/mode/")
        .append(language.name().toLowerCase())
        .append("')");
    loadJs(sb.toString());
  }

  public boolean hasUndo() {
    return webInterface.hasUndo;
  }

  public boolean hasRedo() {
    return webInterface.hasRedo;
  }

  public void setText(String text) {
    String escapedText = text.replace("\n", "\\n").replace("\r", "\\r").replace("'", "\\'");
    loadJs("editor.session.setValue('" + escapedText + "')");
  }

  public String getText() {
    return webInterface.value;
  }

  public void setFontSize(int fontSizeInPx) {
    loadJs("editor.setFontSize(" + fontSizeInPx + ")");
  }

  public void getFontSize(OnJsReturnValueCallback callback) {
    loadJs("editor.getFontSize()", callback);
  }

  public void setShowPrintMargin(boolean showPrintMargin) {
    loadJs("editor.setShowPrintMargin(" + showPrintMargin + ");");
  }

  public void setHighlightActiveLine(boolean highlightActiveLine) {
    loadJs("editor.setHighlightActiveLine(" + highlightActiveLine + ");");
  }

  public void setReadOnly(boolean readOnly) {
    loadJs("editor.setReadOnly(" + readOnly + ");");
  }

  public void setUseWrapMode(boolean useWrapMode) {
    loadJs("editor.getSession().setUseWrapMode(" + useWrapMode + ");");
  }

  public void setUseSoftTabs(boolean useSoftTabs) {
    loadJs("editor.getSession().setUseSoftTabs(" + useSoftTabs + ");");
  }

  public void setTabSize(int tabSize) {
    loadJs("editor.getSession().setTabSize(" + tabSize + ");");
  }

  public void gotoLine(int lineNumber) {
    loadJs("editor.gotoLine(" + lineNumber + ");");
  }

  public void getTotalLines(OnJsReturnValueCallback callback) {
    loadJs("editor.session.getLength()", callback);
  }

  public void insert(String text) {
    loadJs("editor.insert(\"" + text + "\");");
  }

  public void setLanguageFromFile(String filePath) {
    loadJs("setLanguageFromFile(\"" + filePath + "\");");
  }

  public void moveCursorTo(int row, int column) {
    loadJs("editor.session.getSelection().moveCursorTo(" + row + ", " + column + ");");
  }

  public void undo() {
    loadJs("editor.session.getUndoManager().undo()");
  }

  public void redo() {
    loadJs("editor.session.getUndoManager().redo()");
  }

  @Override
  public boolean dispatchKeyEvent(KeyEvent event) {
    return super.dispatchKeyEvent(event);
  }

  @Override
  public InputConnection onCreateInputConnection(EditorInfo outAttr) {
    // this is needed for #dispatchKeyEvent() to be notified.
    return new BaseInputConnection(this, false);
  }

  public void loadJs(String js, OnJsReturnValueCallback callback) {
    evaluateJavascript(
        js,
        (value) -> {
          if (callback != null) callback.onReturnValue(value);
        });
  }

  public void loadJs(String js) {
    loadJs(js, null);
  }

  public void setOnEditorLoadedListener(OnEditorLoadedListener onEditorLoadedListener) {
    this.onEditorLoadedListener = onEditorLoadedListener;
  }

  public EditorOptions getOptions() {
    return this.options;
  }

  public void setOptions(EditorOptions options) {
    this.options = options;
    loadOptions();
  }

  public File getFile() {
    return this.file;
  }

  public void setFile(File file) {
    this.file = file;
  }

  public boolean isModified() {
    return this.modified;
  }

  public void setModified(boolean modified) {
    this.modified = modified;
  }
}

File: editor-ace/src/main/java/com/raredev/vcspace/editor/AceCodeEditor.kt
package com.raredev.vcspace.editor

import android.content.Context
import android.util.AttributeSet
import android.view.KeyEvent
import android.view.inputmethod.BaseInputConnection
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputConnection
import android.webkit.WebView
import com.raredev.vcspace.editor.ace.OnEditorLoadedListener
import com.raredev.vcspace.editor.ace.OnJsReturnValueCallback
import com.raredev.vcspace.editor.ace.Point
import com.raredev.vcspace.editor.ace.VCSpaceEditorLanguage
import com.raredev.vcspace.editor.ace.VCSpaceEditorTheme
import com.raredev.vcspace.editor.ace.VCSpaceWebChromeClient
import com.raredev.vcspace.editor.ace.VCSpaceWebInterface
import com.raredev.vcspace.editor.ace.VCSpaceWebViewClient
import com.raredev.vcspace.editor.ace.options.EditorOptions
import com.raredev.vcspace.editor.ace.options.SearchOptions
import java.io.File

class AceCodeEditor : WebView {
  constructor(context: Context) : super(context)

  constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

  constructor(
      context: Context,
      attrs: AttributeSet?,
      defStyleAttr: Int
  ) : super(context, attrs, defStyleAttr)

  constructor(
      context: Context,
      attrs: AttributeSet?,
      defStyleAttr: Int,
      defStyleRes: Int
  ) : super(context, attrs, defStyleAttr, defStyleRes)

  var onEditorLoadedListener: OnEditorLoadedListener? = null

  var webInterface: VCSpaceWebInterface
  var options: EditorOptions
  var isModified: Boolean = false
  var file: File? = null

  init {
    options = EditorOptions.instance
    webInterface = VCSpaceWebInterface(this)
    addJavascriptInterface(webInterface, "VCSpace")
    webChromeClient = VCSpaceWebChromeClient()
    webViewClient = VCSpaceWebViewClient(this)

    settings.javaScriptEnabled = true
    loadUrl("file:///android_asset/editor/ace-editor/editor.html")
  }

  fun release() {
    destroyEditor()
    destroy()
  }

  fun loadOptions() {
    val sb =
        StringBuilder().apply {
          append("editor.setOptions({")
          append("highlightActiveLine: ${options.highlightActiveLine},")
          append("highlightSelectedWord: ${options.highlightSelectedWord},")
          append("readOnly: ${options.readOnly},")
          append("enableAutoIndent: ${options.enableAutoIndent},")
          append("enableBasicAutocompletion: ${options.enableBasicAutocompletion},")
          append("enableLiveAutocompletion: ${options.enableLiveAutocompletion},")
          append("enableSnippets: ${options.enableSnippets},")
          append("hScrollBarAlwaysVisible: ${options.hScrollBarAlwaysVisible},")
          append("vScrollBarAlwaysVisible: ${options.vScrollBarAlwaysVisible},")
          append("highlightGutterLine: ${options.highlightGutterLine},")
          append("animatedScroll: ${options.animatedScroll},")
          append("showInvisibles: ${options.showInvisibles},")
          append("showPrintMargin: ${options.showPrintMargin},")
          append("fadeFoldWidgets: ${options.fadeFoldWidgets},")
          append("showFoldWidgets: ${options.showFoldWidgets},")
          append("showLineNumbers: ${options.showLineNumbers},")
          append("showGutter: ${options.showGutter},")
          append("displayIndentGuides: ${options.displayIndentGuides},")
          append("fixedWidthGutter: ${options.fixedWidthGutter},")
          append("useSvgGutterIcons: ${options.useSvgGutterIcons},")
          append("useSoftTabs: ${options.useSoftTabs},")
          append("enableEmmet: ${options.enableEmmet},")
          append(
              "mergeUndoDeltas: ${when (options.mergeUndoDeltas) {
        EditorOptions.MergeUndoDeltas.NEVER -> false
        EditorOptions.MergeUndoDeltas.TIMED -> true
        else -> "'always'"
        }},")
          append("selectionStyle: \"${options.selectionStyle.name.lowercase()}\",")
          append("cursorStyle: \"${options.cursorStyle.name.lowercase()}\",")
          append("theme: \"ace/theme/${options.theme.name.lowercase()}\",")
          append("mode: \"ace/mode/${options.language.name.lowercase()}\",")
          append("newLineMode: \"${options.newLineMode.name.lowercase()}\",")
          append("foldStyle: \"${options.foldStyle.name.lowercase()}\",")
          append("firstLineNumber: ${options.firstLineNumber},")
          append("fontSize: ${options.fontSize},")
          append("printMarginColumn: ${options.printMarginColumn},")
          append("tabSize: ${options.tabSize},")
          append("maxLines: ${if (options.maxLines == 0) "undefined" else options.maxLines},")
          append("minLines: ${if (options.minLines == 0) "undefined" else options.minLines},")
          append("scrollPastEnd: ${options.scrollPastEnd}")
          append("})")
        }
    loadJs(sb.toString())
  }

  fun setTheme(theme: VCSpaceEditorTheme) {
    loadJs("editor.setTheme('ace/theme/${theme.name.lowercase()}')")
  }

  fun setLanguage(language: VCSpaceEditorLanguage) {
    loadJs("editor.session.setMode('ace/mode/${language.name.lowercase()}')")
  }

  fun hasUndo() = webInterface.hasUndo

  fun hasRedo() = webInterface.hasRedo

  fun setText(text: String) =
      loadJs(
          "editor.session.setValue('${text.replace("\n", "\\n").replace("\r", "\\r").replace("'", "\\'")}')")

  fun getText() = webInterface.value

  fun setFontSize(fontSizeInPx: Int) = loadJs("editor.setFontSize($fontSizeInPx)")

  fun getFontSize(callback: OnJsReturnValueCallback) = loadJs("editor.getFontSize()", callback)

  fun setShowPrintMargin(showPrintMargin: Boolean) =
      loadJs("editor.setShowPrintMargin($showPrintMargin);")

  fun setHighlightActiveLine(highlightActiveLine: Boolean) =
      loadJs("editor.setHighlightActiveLine($highlightActiveLine);")

  fun setReadOnly(readOnly: Boolean) = loadJs("editor.setReadOnly($readOnly);")

  fun setUseWrapMode(useWrapMode: Boolean) = loadJs("editor.session.setUseWrapMode($useWrapMode);")

  fun setUseSoftTabs(useSoftTabs: Boolean) = loadJs("editor.session.setUseSoftTabs($useSoftTabs);")

  fun setTabSize(tabSize: Int) = loadJs("editor.session.setTabSize($tabSize);")

  fun gotoLine(lineNumber: Int) = loadJs("editor.gotoLine($lineNumber);")

  fun getTotalLines(callback: OnJsReturnValueCallback) =
      loadJs("editor.session.getLength()", callback)

  fun insert(text: String) =
      loadJs(
          "editor.insert(\"${text.replace("\n", "\\n").replace("\r", "\\r").replace("'", "\\'")}\");")

  fun setLanguageFromFile(filePath: String?) = loadJs("setLanguageFromFile(\"$filePath\");")

  fun undo() = loadJs("editor.session.getUndoManager().undo()")

  fun redo() = loadJs("editor.session.getUndoManager().redo()")

  fun destroyEditor() = loadJs("editor.session.destroy()")

  fun alignCursors() = loadJs("editor.alignCursors()")

  fun applyComposition(text: Any?, composition: Any?) =
      loadJs("editor.applyComposition($text, $composition)")

  fun autoIndent() = loadJs("editor.autoIndent()")

  fun blockIndent() = loadJs("editor.blockIndent()")

  fun blockOutdent() = loadJs("editor.blockOutdent()")

  fun blur() = loadJs("editor.blur()")

  fun centerSelection() = loadJs("editor.centerSelection()")

  fun copyLinesDown() = loadJs("editor.copyLinesDown()")

  fun copyLinesUp() = loadJs("editor.copyLinesUp()")

  fun duplicateSelection() = loadJs("editor.duplicateSelection()")
  // fun endOperation(e: Any? = null) = loadJs("editor.endOperation()")

  fun find(needle: String, options: SearchOptions? = null, animate: Boolean = true) {
    val optionsString =
        options?.let {
          val parts = mutableListOf<String>()
          with(it) {
            if (preventScroll) parts += "preventScroll: true"
            if (backwards) parts += "backwards: true"
            // TODO:
            // start?.let {
            //   parts += "start: {start: {row: ${it.start.row}, column: ${it.start.column}}, end:
            // {row: ${it.end.row}, column: ${it.end.column}}}"
            // }
            // range?.let {
            //   parts += "range: {start: {row: ${it.start.row}, column: ${it.start.column}}, end:
            // {row: ${it.end.row}, column: ${it.end.column}}}"
            // }
            if (skipCurrent) parts += "skipCurrent: true"
            if (preserveCase) parts += "preserveCase: true"
            if (regExp) parts += "regExp: true"
            if (wholeWord) parts += "wholeWord: true"
            if (caseSensitive) parts += "caseSensitive: true"
            if (wrap) parts += "wrap: true"
          }
          if (parts.isNotEmpty()) ", {${parts.joinToString(", ")}}" else ""
        } ?: ""

    val animateString = if (animate) "true" else "false"
    val js = "editor.find('$needle'$optionsString, $animateString)"
    loadJs(js)
  }

  fun findAll(needle: String, options: SearchOptions? = null, additive: Boolean = false) {
    // TODO: Implement this method
  }

  fun findLinkAt(row: Int, column: Int) = loadJs("editor.findLinkAt($row, $column)")

  fun findNext(options: SearchOptions? = null, animate: Boolean = false) {
    // TODO: Implement this method
  }

  fun findPrevious(options: SearchOptions? = null, animate: Boolean = false) {
    // TODO: Implement this method
  }

  fun focus() = loadJs("editor.focus()")

  fun getCursorPosition(): Point? = webInterface.cursorPosition

  fun jumpToMatching(select: Boolean, expand: Boolean) =
      loadJs("editor.jumpToMatching($select, $expand")

  fun moveCursorTo(row: Int, column: Int) = loadJs("editor.moveCursorTo($row, $column)")

  fun moveCursorToPosition(pos: Point): Unit = moveCursorTo(pos.row, pos.column)

  fun moveLinesDown() = loadJs("editor.moveLinesDown()")

  fun moveLinesUp() = loadJs("editor.moveLinesUp()")

  fun navigateFileEnd() = loadJs("editor.navigateFileEnd()")

  fun navigateFileStart() = loadJs("editor.navigateFileStart()")

  fun navigateLineEnd() = loadJs("editor.navigateLineEnd()")

  fun navigateLineStart() = loadJs("editor.navigateLineStart()")

  fun replace(replacement: String, options: SearchOptions? = null) {
    // TODO: Implement this method
  }

  fun replaceAll(replacement: String, options: SearchOptions? = null) {
    // TODO: Implement this method
  }

  fun resize(force: Boolean = false) = loadJs("editor.resize($force)")

  fun selectAll() = loadJs("editor.selectAll()")

  fun showKeyboardShortcuts() = loadJs("editor.showKeyboardShortcuts()")

  fun showSettingsMenu() = loadJs("editor.showSettingsMenu()")

  fun toggleBlockComment() = loadJs("editor.toggleBlockComment()")

  fun loadJs(js: String, callback: OnJsReturnValueCallback? = null) {
    evaluateJavascript(js) { value -> callback?.onReturnValue(value) }
  }

  override fun dispatchKeyEvent(event: KeyEvent): Boolean {
    return super.dispatchKeyEvent(event)
  }

  override fun onCreateInputConnection(outAttrs: EditorInfo): InputConnection {
    // this is needed for #dispatchKeyEvent() to be notified.
    return BaseInputConnection(this, false)
  }
}
